library/core/src/intrinsics/simd.rs

```diff
@@ -0,0 +1,412 @@
+
+    /// Add elements within a vector from left to right.
+    ///
+    /// Starting with the value `y`, add the elements of `x` and accumulate.
+    ///
+    /// `T` must be a vector of integer or floating-point primitive types.
+    ///
+    /// `U` must be the element type of `T`.
+    pub fn simd_reduce_add_ordered<T, U>(x: T, y: U) -> U;
+
+    /// Multiply elements within a vector from left to right.
+    ///
+    /// Starting with the value `y`, multiply the elements of `x` and accumulate.
+    ///
+    /// `T` must be a vector of integer or floating-point primitive types.
+    ///
+    /// `U` must be the element type of `T`.
+    pub fn simd_reduce_mul_ordered<T, U>(x: T, y: U) -> U;
+
+    /// Check if all mask values are true.
+    ///
+    /// `T` must be a vector of integer primitive types.
+    ///
+    /// # Safety
+    /// `x` must contain only `0` or `!0`.
+    pub fn simd_reduce_all<T>(x: T) -> bool;
+
+    /// Check if all mask values are true.
+    ///
+    /// `T` must be a vector of integer primitive types.
+    ///
+    /// # Safety
+    /// `x` must contain only `0` or `!0`.
+    pub fn simd_reduce_any<T>(x: T) -> bool;
+
+    /// Return the maximum element of a vector.
+    ///
+    /// For floating-point values, uses IEEE-754 `maxNum`.
+    ///
+    /// `T` must be a vector of integer or floating-point primitive types.
+    ///
+    /// `U` must be the element type of `T`.
+    pub fn simd_reduce_max<T, U>(x: T) -> U;
+
+    /// Return the minimum element of a vector.
+    ///
+    /// For floating-point values, uses IEEE-754 `minNum`.
+    ///
+    /// `T` must be a vector of integer or floating-point primitive types.
+    ///
+    /// `U` must be the element type of `T`.
+    pub fn simd_reduce_min<T, U>(x: T) -> U;
+
+    /// Logical "and" all elements together.
+    ///
+    /// `T` must be a vector of integer or floating-point primitive types.
+    ///
+    /// `U` must be the element type of `T`.
+    pub fn simd_reduce_and<T, U>(x: T) -> U;
+
+    /// Logical "or" all elements together.
+    ///
+    /// `T` must be a vector of integer or floating-point primitive types.
+    ///
+    /// `U` must be the element type of `T`.
+    pub fn simd_reduce_or<T, U>(x: T) -> U;
+
+    /// Logical "exclusive or" all elements together.
+    ///
+    /// `T` must be a vector of integer or floating-point primitive types.
+    ///
+    /// `U` must be the element type of `T`.
+    pub fn simd_reduce_xor<T, U>(x: T) -> U;
+
+    /// Truncate an integer vector to a bitmask.
+    ///
+    /// Each element is truncated to a single bit and packed into the result.
+    ///
+    /// The bit order depends on the byte endianness.
+    /// The bitmask is always packed into the smallest/first bits, but the order is LSB-first for
+    /// little endian and MSB-first for big endian.
+    /// In other words, the LSB corresponds to the first vector element for little endian,
+    /// and the last vector element for big endian.
```

Comment:
I have no idea what this means.^^ There's too many different notions of "first" here. Also, there's two cases to consider (output a packed int vs output an array).

What about something like this:
No matter whether the output is an array or an unsigned integer, it is treated as a single contiguous list of bits. The bitmask is always packed on the least-significant side of the output, and padded with 0s in the most-significant bits. The order of the bits depends on endianess:
- On little endian, the least significant bit corresponds to the *first* vector element.
- On big endian, the least significant bit corresponds to the *last* vector element.

I think this also needs examples to have any chance of being comprehensible.

As always, please first state the types and then start discussing the details; without knowing what `U` is the rest of this is even harder to understand.

Reply:
This is much better. I took your text and added a simple example